bson/src/main/org/bson/codecs/pojo/ConventionObjectIdGeneratorsImpl.java

```diff
@@ -30,6 +31,9 @@ public void apply(final ClassModelBuilder<?> classModelBuilder) {
                    classModelBuilder.idGenerator(IdGenerators.OBJECT_ID_GENERATOR);
                } else if (classModelBuilder.getIdGenerator() == null && idType.equals(BsonObjectId.class)) {
                    classModelBuilder.idGenerator(IdGenerators.BSON_OBJECT_ID_GENERATOR);
+               } else if (classModelBuilder.getIdGenerator() == null && idType.equals(String.class)
```

Comment:
This seems like a design flaw.  Elsewhere (i.e. PojoCodecImpl) we do not make assumptions about which Codecs support changing BSON representation, but now here we are assuming that whatever Codec happens to be registered for String.class does support it (it's true out default one does, but it could be changed by the application), and no others are (a user could register their own for some other type).

But given that we defer Codec resolution until first use, I don't see a way out of this.  

Any ideas?

Reply:
I don't really see why this is a problem... This condition will only be true if the user has explicitly marked their String id field with `@BsonRepresentation(BsonType.ObjectId)`. By doing this, they are implying that whatever codec they're using for `Strings` supports changing the BSON rep. 

If they use the annotation and their codec does not support changing the BSON rep, an exception will be thrown anyway when the String codec is resolved the first time.

Reply:
> we are assuming that whatever Codec happens to be registered for String.class does support it

I see your point.  This isn't really a problem

> and no others are

I was thinking of an application-defined class, e.g. 

     public class Guid {
           private final String guid;
     }

but I don't see how we could create an Id generator that populates this anyway.  The application would have to register it as a convention explicitly.

So I think this is ok.  False alarm.